lib/gears/shape.lua

```diff
@@ -41,14 +41,128 @@
 -- @themelib gears.shape
 ---------------------------------------------------------------------------
 local g_matrix = require( "gears.matrix" )
+local g_math   = require( "gears.math" )
 local unpack   = unpack or table.unpack -- luacheck: globals unpack (compatibility with Lua 5.1)
 local atan2    = math.atan2 or math.atan -- lua 5.3 compat
 local min      = math.min
 local cos      = math.cos
 local sin      = math.sin
+local pow      = math.pow -- luacheck: globals math.pow (math.pow is deprecated in lua5.3, but is still there)
+local abs      = math.abs
 
 local module = {}
 
+--- Add a squircle shape with only some of the corner are "circled" to the current path
+-- the squircle is not exactly as the definition
+-- it will expand to the shape's width and height
+-- kinda like an ellipse
+--
+-- @DOC_gears_shape_partial_squircle_EXAMPLE@
+--
+-- @param cr A cairo context
+-- @tparam number width The shape width
+-- @tparam number height The height constraint
+-- @tparam boolean tl If the top left corner is rounded
+-- @tparam boolean tr If the top right corner is rounded
+-- @tparam boolean br If the bottom right corner is rounded
+-- @tparam boolean bl If the bottom left corner is rounded
+-- @tparam number rate The "squareness" of the squircle
+-- @tparam number delta The "smoothness" of the shape
+-- @staticfct gears.shape.partial_squircle
+function module.partial_squircle(cr, width, height, tl, tr, br, bl, rate, delta)
+   -- rate ~ 2 can be used by icon
+   -- this shape doesn't really fit clients
+   -- but you can still use with rate ~ 32
+   rate = rate or 2
+   -- smaller the delta the smoother the shape
+   -- but probably more laggy
+   delta = delta or 0.05
+
+   -- just ellipse things
+   local a = width / 2
+   local b = height / 2
+   local phi = 0
+
+   -- move to origin
+   -- the shape goes counter clock wise
+   -- start from (w h / 2)
+   cr:save()
+   cr:translate(a, b)
+   cr:move_to(a, 0)
+
+   -- draw the corner if that corner is rounded
+   local curved_corner = function()
+      local end_angle = phi + math.pi / 2
+      while phi < end_angle do
+         local cosphi = cos(phi)
+         local sinphi = sin(phi)
+         local x = a * pow(abs(cosphi), 1 / rate) * g_math.sign(cosphi)
+         local y = b * pow(abs(sinphi), 1 / rate) * g_math.sign(sinphi)
+         -- so weird, y axis is inverted
+         cr:line_to(x, -y)
+         phi = phi + delta
+      end
```

Comment:
Hm... if someone chooses `delta` small enough, this could turn into an endless loop if `phi == phi + delta` due to rounding issues. Should this shape somehow protect against that? One quick idea is to assert that `delta > 0`. A second quick idea would be to instead count integers (something like `for i = 1, n do phi = math.pi / 2 * i / n`).

Does any of this make sense?

Reply:
Github doesn't notify me when there's a comment so I didn't see this after reloading the page. 

Sure, asserting `delta > 0` sounds great. The for loop will work too, but is doing `phi = math.pi / 2 * i / n` is more works than just adding it by `delta`?

Reply:
Dunno if it really makes a difference. I guess not. But I would expect `2*pi * i / n` to have less error. Adding a small value many times magnifies rounding errors. Which is not really important here since `phi` is just used to calculate the next point based on an analytic formula....

Feel free to ignore me here.

Reply:
Wait, that might actually valid. Since we're doing `sin(...)^(1/rate)` the error will even be amplified. Maybe I'll decide after some math and ploting first

Reply:
On second thought, like you said, the `phi` is just the angle of the next point in polar coord, and since the shape is kinda continuous, this doesn't really matter.

Reply:
For the above reason, plus this one, I decided to keep using the `while` loop. Since we only know how small delta should not be and make an assertion. We don't know how many strokes at least is needed to draw the shape so we can't make the assertion with the `for` loop.